File: android/src/main/java/com/reactnativeimagecolors/ImageColorsModule.java
package com.reactnativeimagecolors;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.util.Base64;

import androidx.annotation.NonNull;
import androidx.palette.graphics.Palette;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.facebook.react.bridge.WritableMap;

import java.net.MalformedURLException;
import java.net.URI;
import java.net.URLConnection;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

public class ImageColorsModule extends ReactContextBaseJavaModule {
  private static final String base64Scheme = "data";
  private final ExecutorService executorService;
  private Integer pixelSpacing;


    ImageColorsModule(ReactApplicationContext reactContext) {
      super(reactContext);
      executorService = new ThreadPoolExecutor(
        0,
        Integer.MAX_VALUE,
        30L,
        TimeUnit.SECONDS,
        new SynchronousQueue<>()
      );
    }

    @NonNull
    @Override
    public String getName() {
        return "ImageColors";
    }


    /**
     * https://gist.github.com/maxjvh/a6ab15cbba9c82a5065d
     * pixelSpacing tells how many pixels to skip each pixel.
     * If pixelSpacing > 1: the average color is an estimate, but higher values mean better performance.
     * If pixelSpacing == 1: the average color will be the real average.
     * If pixelSpacing < 1: the method will most likely crash (don't use values below 1).
     */
    private int calculateAverageColor(@NonNull Bitmap bitmap) {
        int R = 0;
        int G = 0;
        int B = 0;
        int height = bitmap.getHeight();
        int width = bitmap.getWidth();
        int n = 0;
        int[] pixels = new int[width * height];

        bitmap.getPixels(pixels, 0, width, 0, 0, width, height);

        int spacing = 5;

        if(pixelSpacing != null){
            spacing = pixelSpacing;
        }

        for (int i = 0; i < pixels.length; i += spacing) {
            int color = pixels[i];
            R += Color.red(color);
            G += Color.green(color);
            B += Color.blue(color);
            n++;
        }
        return Color.rgb(R / n, G / n, B / n);
    }

    private String getHex(int rgb) {
        return String.format("#%06X", (0xFFFFFF & rgb));
    }

    private int parseColorFromHex(String colorHex) {
        return Color.parseColor(colorHex);
    }

    @ReactMethod
    public void getColors(String source, ReadableMap config, Promise promise) {
      executorService.execute(() -> {
          try {
              String defColor = "#000000";
              pixelSpacing = null;

              if (config != null){
                  if (config.hasKey("defaultColor")) {
                      defColor = config.getString("defaultColor");
                  }
                  if (config.hasKey("pixelSpacing")) {
                      pixelSpacing = config.getInt("pixelSpacing");
                  }
              }

              int defColorInt = parseColorFromHex(defColor);

              WritableMap resultMap = Arguments.createMap();
              resultMap.putString("platform", "android");


              Context context = getReactApplicationContext();
              int resourceId = context.getResources().getIdentifier(source, "drawable", context.getPackageName());
              Bitmap image = null;

              if (resourceId == 0) {
                  // resource is not a local file
                  // could be a URL, base64.
                  URI uri = new URI(source);
                  String scheme = uri.getScheme();

                  if(scheme == null) throw new Exception("Invalid URI scheme");

                  if (scheme.equals(base64Scheme)) {
                      String[] parts = source.split(",");
                      String base64Uri = parts[1];
                      byte[] decodedString = Base64.decode(base64Uri, Base64.DEFAULT);
                      image = BitmapFactory.decodeByteArray(decodedString, 0, decodedString.length);
                  } else {
                      URLConnection urlConnection = uri.toURL().openConnection();

                      ReadableMap headers = config != null ? config.getMap("headers") : null;
                      if(headers != null){
                          ReadableMapKeySetIterator iterator = headers.keySetIterator();
                          while(iterator.hasNextKey()){
                              String key = iterator.nextKey();
                              urlConnection.setRequestProperty(key, headers.getString(key));
                          }
                      }

                      image = BitmapFactory.decodeStream(urlConnection.getInputStream());
                  }
              } else {
                  image = BitmapFactory.decodeResource(context.getResources(), resourceId);
              }

              if (image == null) throw new Exception("Invalid image URI – failed to get image");

              int rgbAvg = calculateAverageColor(image);
              String hexAvg = getHex(rgbAvg);
              resultMap.putString("average", hexAvg);

                  Palette.Builder builder = new Palette.Builder(image);
                  builder.generate(palette -> {
                      try {
                          if (palette != null) {
                                  int rgb = palette.getDominantColor(defColorInt);
                                  String hex = getHex(rgb);
                                  resultMap.putString("dominant", hex);

                                  int rgb1 = palette.getVibrantColor(defColorInt);
                                  String hex1 = getHex(rgb1);
                                  resultMap.putString("vibrant", hex1);

                                  int rgb2 = palette.getDarkVibrantColor(defColorInt);
                                  String hex2 = getHex(rgb2);
                                  resultMap.putString("darkVibrant", hex2);

                                  int rgb3 = palette.getLightVibrantColor(defColorInt);
                                  String hex3 = getHex(rgb3);
                                  resultMap.putString("lightVibrant", hex3);

                                  int rgb4 = palette.getDarkMutedColor(defColorInt);
                                  String hex4 = getHex(rgb4);
                                  resultMap.putString("darkMuted", hex4);

                                  int rgb5 = palette.getLightMutedColor(defColorInt);
                                  String hex5 = getHex(rgb5);
                                  resultMap.putString("lightMuted", hex5);

                                  int rgb6 = palette.getMutedColor(defColorInt);
                                  String hex6 = getHex(rgb6);
                                  resultMap.putString("muted", hex6);

                                  promise.resolve(resultMap);
                          } else {
                              throw new Exception("Palette was null");
                          }
                      } catch (Exception e) {
                          handleException(e, promise);
                      }
                  });

          } catch (MalformedURLException e) {
              handleException(new Exception("Invalid URL"), promise);
          } catch (Exception e) {
              handleException(e, promise);
          }
      });
    }


    private void handleException(Exception e, Promise promise) {
        e.printStackTrace();
        promise.reject("Error", "ImageColors: " + e.getMessage());
    }
}

File: android/src/main/java/com/reactnativeimagecolors/ImageColorsModule.kt
package com.reactnativeimagecolors

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.util.Base64
import android.util.Log
import android.webkit.URLUtil
import androidx.palette.graphics.Palette

import expo.modules.core.errors.ModuleDestroyedException
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

import kotlinx.coroutines.*
import kotlin.math.ceil

import java.net.MalformedURLException
import java.net.URI

class Config : Record {
  @Field
  val fallback: String = "#000000"

  @Field
  val headers: Map<String, String>? = null

  @Field
  val pixelSpacing: Int = 5
}

class ImageColorsModule : Module() {
  private val service = CoroutineScope(Dispatchers.IO)

  /**
   * pixelSpacing tells how many pixels to skip each pixel.
   * If pixelSpacing > 1: the average color is an estimate, but higher values mean better performance.
   * If pixelSpacing == 1: the average color will be the real average.
   * If pixelSpacing < 1: the method will most likely crash (don't use values below 1).
   */
  private fun calculateAverageColor(bitmap: Bitmap, pixelSpacing: Int): Int {
    val segmentWidth = 500

    val width = bitmap.width
    val height = bitmap.height

    val numSegments = ceil(width.toDouble() / segmentWidth).toInt()
    val segmentPixels = IntArray(segmentWidth * height)

    var redSum = 0
    var greenSum = 0
    var blueSum = 0
    var pixelCount = 0

    for (i in 0 until numSegments) {
      val xStart = i * segmentWidth
      val xEnd = minOf(width, (i + 1) * segmentWidth)

      bitmap.getPixels(segmentPixels, 0, segmentWidth, xStart, 0, xEnd - xStart, height)

      for (index in segmentPixels.indices step pixelSpacing) {
        redSum += Color.red(segmentPixels[index])
        greenSum += Color.green(segmentPixels[index])
        blueSum += Color.blue(segmentPixels[index])
        pixelCount++
      }
    }
    return if (pixelCount == 0) {
      Color.BLACK
    } else {
      val red = redSum / pixelCount
      val green = greenSum / pixelCount
      val blue = blueSum / pixelCount
      Color.rgb(red, green, blue)
    }
  }

  private fun parseFallbackColor(hex: String): String {
    if(!hex.matches(Regex("^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$"))) {
      throw Exception("Invalid fallback hex color. Must be in the format #ffffff or #fff")
    }

    if(hex.length == 7) {
      return hex
    }

    return "#${hex[1]}${hex[1]}${hex[2]}${hex[2]}${hex[3]}${hex[3]}"
  }

  private fun getHex(rgb: Int): String {
    return String.format("#%06X", 0xFFFFFF and rgb)
  }

  private fun handleError(promise: Promise, err: Exception) {
    promise.reject("[ImageColors]", err.message, err)
  }

  override fun definition() = ModuleDefinition {
    Name("ImageColors")

    AsyncFunction("getColors") { uri: String, config: Config, promise: Promise ->
      service.launch {
        try {
          val fallbackColor = parseFallbackColor(config.fallback)
          val fallbackColorInt = Color.parseColor(fallbackColor)
          var image: Bitmap? = null

          val context = appContext.reactContext
          val resourceId =
            context?.resources?.getIdentifier(uri, "drawable", context.packageName) ?: 0

          // check if local resource
          if (context != null && resourceId != 0) {
            image = BitmapFactory.decodeResource(context.resources, resourceId)
          }

          // check if base64
          if (uri.startsWith("data:image")) {
            val base64Uri = uri.split(",")[1]
            val decodedBytes = Base64.decode(base64Uri, Base64.DEFAULT)

            image = BitmapFactory.decodeByteArray(decodedBytes, 0, decodedBytes.size)
          }

          if (URLUtil.isValidUrl(uri)) {
            val parsedUri = URI(uri)
            val connection = parsedUri.toURL().openConnection()

            if (config.headers != null) {
              for (header in config.headers) {
                connection.setRequestProperty(header.key, header.value)
              }
            }

            image = BitmapFactory.decodeStream(connection.getInputStream())
          }

          if (image == null) {
            throw Exception("Filed to get image")
          }

          val paletteBuilder = Palette.Builder(image)
          val result: MutableMap<String, String> = mutableMapOf()

          result["average"] = getHex(calculateAverageColor(image, config.pixelSpacing))
          result["platform"] = "android"

          try {
            val palette = paletteBuilder.generate()

            result["dominant"] = getHex(palette.getDominantColor(fallbackColorInt))
            result["vibrant"] = getHex(palette.getVibrantColor(fallbackColorInt))
            result["darkVibrant"] = getHex(palette.getDarkVibrantColor(fallbackColorInt))
            result["lightVibrant"] = getHex(palette.getLightVibrantColor(fallbackColorInt))
            result["muted"] = getHex(palette.getMutedColor(fallbackColorInt))
            result["darkMuted"] = getHex(palette.getDarkMutedColor(fallbackColorInt))
            result["lightMuted"] = getHex(palette.getLightMutedColor(fallbackColorInt))

            promise.resolve(result)
          } catch (err: Exception) {
            result["dominant"] = fallbackColor
            result["vibrant"] = fallbackColor
            result["darkVibrant"] = fallbackColor
            result["lightVibrant"] = fallbackColor
            result["muted"] = fallbackColor
            result["darkMuted"] = fallbackColor
            result["lightMuted"] = fallbackColor

            promise.resolve(result)
          }
        } catch (err: MalformedURLException) {
          handleError(promise, Exception("Invalid URL"))
        } catch (err: Exception) {
          handleError(promise, err)
        }
      }

      OnDestroy {
        try {
          service.cancel(ModuleDestroyedException())
        } catch (e: IllegalStateException) {
          Log.e("[ImageColors]", "The scope does not have a job in it")
        }
      }
    }
  }
}
